// blueprint/blueprint.js
var blueprint_utils = {
  get_blueprint: function(id, err, blueprint){
    Bp.findOne({ id: id }, function(err_, bp) {
      if (err_){
        err(err_);
      }else{
        blueprint(bp);
      }
    });
  },
  create_blueprint: function(instance_id, tools, defect_tracker, auth, projects, documentation, err, result){
    Bp.create({
      id: instance_id,
      tools: tools,
      defect_tracker: defect_tracker,
      auth: auth,
      projects: projects,
      documentation: documentation
    }).done(function(err_, bp) {
      // Error handling
      if (err_) {
        err(err_);
      // The Blueprint was created successfully!
      }else {
        //Blueprint created
        result(bp);
      }
    });
  },
  get_kit_blueprint: function(err, blueprint){
    var config =  require('config/config.js').get_config_data();
    if(config  instanceof Error){
      err('Error reading the configuration file.');
    }else{
      blueprint_utils.get_blueprint(config.id, function(errbp){
        if(errbp){
          err(errbp);
        }
      }, function(bp){
        if(bp === undefined){
          //If the blueprint does not exist in our model that means we haven't created the record in our model
          blueprint_utils.create_blueprint(config.id, config.tools, config.defect_tracker, config.auth, config.projects, config.documentation, function(errc){
            err(errc);
          }, function(bpc){
            blueprint(bpc);
          });
        }else{
          blueprint(bp);
        }
      });
    }
  }
}

module.exports = blueprint_utils;
