// kit-ops/kit-ops.js
var config_yaml = require('config-yaml/config-yaml').get_config_yaml();
var http = require('http');
module.exports = {
  get_opt: function(opt, callback){
    var url = config_yaml.runtime.protocol + '://' + config_yaml.runtime.ip + ':' + config_yaml.runtime.port+ config_yaml.runtime.resource_uri + 'option/' + opt;
    sails.log.debug('get_opt: url='+url);
    var reqGet = http.get(url, function(res) {
      res.on('data', function (chunk) {
        chunk = JSON.parse(chunk)[0];
        sails.log.debug('get_opt: result: '+chunk);
        callback(null, chunk);
      });
      
      res.on('error', function(e){
        console.error('Error get_opt:'+e.message);
        callback('Error get_blueprint_id:'+e.message, null);
      });
    });
    
    reqGet.end();
    reqGet.on('error', function (e) {
      console.error('Error closing the request get_opt:'+e.message);
      callback('Error closing the request get_opt:'+e.message, null);
    });
  },
  create_kit_admin: function(email, callback){
    post_options = {
            host: config_yaml.runtime.ip,
            port: config_yaml.runtime.port,
            path: config_yaml.runtime.resource_uri + 'option/kit_admin/'+email,
            method: 'POST',
            rejectUnauthorized: false,
            requestCert: true,
            agent: false,
            headers: {
              'accept': 'application/json',
              'content-type': 'application/json'
            }
        };
      var payload = { optname: 'kit_admin', optvalue: email };
    
      var post_req = http.request(post_options, function(res) {
          res.setEncoding('utf8');
          res.on('data', function (chunk) {
            chunk = JSON.parse(chunk);
            if(chunk.affectedRows == 1){
              callback(null, true);
            }else{
              callback(null, false);
            }
          });
      });
      
      post_req.on('error', function (e) {
          console.log('Kit Registration API Error: ' + e.message);
          callback(e.message, null);
      });
      
      post_req.write(JSON.stringify(payload));
      post_req.end();
  },
  is_admin: function(email, callback){
    var url = config_yaml.runtime.protocol + '://' + config_yaml.runtime.ip + ':' + config_yaml.runtime.port+ config_yaml.runtime.resource_uri + 'option/kit_admin';
    sails.log.debug('is_admin: url='+url);
    var reqGet = http.get(url, function(res) {
      res.on('data', function (chunk) {
        var admin_found = false;
        if(res.statusCode == 200){
          JSON.parse(chunk).forEach(function(admin){
            if(admin.option_value === email){
              admin_found = true;
              sails.log.debug('is_admin: True');
            }
            else{
              sails.log.debug('is_admin: False');
            }
          });
          callback(null, admin_found);
        }else{
          callback(null, admin_found);
        }
      });
      
      res.on('error', function(e){
        console.error('Error is_admin:'+e.message);
        callback('Error is_admin:'+e.message, null);
      });
    });
    
    reqGet.end();
    reqGet.on('error', function (e) {
      console.error('Error closing the request is_admin:'+e.message);
      callback('Error closing the request is_admin:'+e.message, null);
    });
  },
  kit_has_admin: function(callback){
   
    var url = config_yaml.runtime.protocol + '://' + config_yaml.runtime.ip + ':' + config_yaml.runtime.port+ config_yaml.runtime.resource_uri + 'option/kit_admin';
    sails.log.debug('kit_has_admin: url='+url);
    var reqGet = http.get(url, function(res) {
      res.on('data', function (chunk) {
        if(res.statusCode === 404){
          callback(null, false);
        }else if(res.statusCode === 200){
          chunk = JSON.parse(chunk);
          sails.log.debug('kit_has_admin: result: '+chunk);
          if(chunk.length > 0){
            callback(null, true);
          }else{
            callback(null, false);
          }
        }else{
          callback(null, false);
        }
      });
      
      res.on('error', function(e){
        console.error('Error kit_has_admin:'+e.message);
        callback('Error kit_has_admin:'+e.message, null);
      });
    });
    
    reqGet.end();
    reqGet.on('error', function (e) {
      console.error('Error closing the request kit_has_admin:'+e.message);
      callback('Error closing the request kit_has_admin:'+e.message, null);
    });
  },
  enable_notifications: function(enable, notificationID, callback){
      
    var url = config_yaml.runtime.protocol + '://' + config_yaml.runtime.ip + ':'
      + config_yaml.runtime.port+ config_yaml.runtime.resource_uri
      + 'option/' + notificationID + '/' +enable;

    var reqGet = http.get(url, function(res) {
      res.on('data', function (chunk) {
        if(res.statusCode === 404){
          callback(null, false);
        }else if(res.statusCode === 200){
          chunk = JSON.parse(chunk);
          if(chunk.length > 0){
            callback(null, true);
          }else{
            callback(null, false);
          }
        }else{
          callback(null, false);
        }
      });
      
      res.on('error', function(e){
        console.error('Error enable_notifications:'+e.message);
        callback('Error enable_notifications:'+e.message, null);
      });
    });
    
    reqGet.end();
    reqGet.on('error', function (e) {
      console.error('Error closing the request enable_notifications:'+e.message);
      callback('Error closing the request enable_notifications:'+e.message, null);
    });
  }
}
