// maestro-exec/maestro-exec.js
"use strict";

//var sys = require('sys');
var executable = require('executable');
var exec = require('child_process').exec;

var maestroexec =  module.exports = {

    /* future use  */
    RUN_LOCAL : true,
    /*
    defaults :  {
        env: process.env.NODE_ENV || 'development',
        config: {
            paths: {
                environments: __dirname + '/environments'
            }
        }
    },
    */

    execCmd: function (cmd, params, callback) {
        this.execLocal(cmd, params, callback);
    },
    execLocal: function(cmd, params, callback) {
        try {
            if (executable.sync(cmd)) {
                exec(cmd + ' ' + params, function (error, stdout, stderr) {
                    if (error) {
                        callback(null, error);
                    }
                    else {
                        callback(stdout, null);
                    }
                })
            }
        } catch (e) {
            callback(null, e.message);
        }
    }

};
