// maestro-exec/maestro-exec.js
"use strict";

var executable = require('executable');
var exec = require('child_process').exec,
    child;

var maestroexec =  module.exports = {

    /* future use  */
    RUN_LOCAL : true,
    /*
    defaults :  {
        env: process.env.NODE_ENV || 'development',
        config: {
            paths: {
                environments: __dirname + '/environments'
            }
        }
    },
    */

    execCmd: function(cmd, params, options, callback) {
             this.execLocal(cmd, params, options, callback);
    },

    /*execCmd: function(cmd, params, callback) {
       this.execLocal(cmd, params, function(code, error) {
           (error) ? callback(false) : callback(true);
       });
    },*/

    execLocal: function(cmd, params, options, callback) {
        try {
            if (executable.sync(cmd)) {
                child = exec(((params) ? cmd + ' ' + params : cmd), function (error, stdout, stderr) {
                    if(options && options.status_only) {
                        (error) ? callback(false, null) : callback(true, null);
                    } else {
                        if (error) {
                            callback(null, error);
                        }
                        else {
                            callback(stdout, null);
                        }
                    }
                });
            }
        } catch (e) {
            if(options && options.status_only) {
                callback(false, null);
            } else {
                callback(null, e.message);
            }
        }
    },
    createProject: function(params, callback){
  		var childProcess = require('child_process'),
  		ls;
  		ls = childProcess.exec('/bin/bash /var/lib/forj/newproject.sh '+params, function (error, stdout, stderr) {
  			if(error){
  				console.error('Project creation failed: '+error);
  			}else{
  				console.info('Project Created: '+stdout);
  			}
  		});
  		 ls.on('exit', function (code) {
  				if(code==1){
  						callback(false, null);
  				}else{
  						callback(true, null);
  				}
  		 });
  	}

};
