// maestro-exec/maestro-exec.js
"use strict";

var exec = require('child_process').exec;

var maestroexec =  module.exports = {

    /* future use  */
    RUN_LOCAL : true,


    /*
    cmd: String The command to run, with space-separated arguments       

    callback(error, stdout, stderr):
       On success: error will be null.
       On error: error.code will be the exit code    
    */
    execCmd: function(cmd, callback) {
        exec(cmd, function (error, stdout, stderr) {
            callback(error, stdout, stderr);
        });
    },


    /*
    cmd: String The command to run, with space-separated arguments       

    options: Object;
        cwd String Current working directory of the child process
        env Object Environment key-value pairs
        encoding String (Default: 'utf8')
        timeout Number (Default: 0)
        maxBuffer Number (Default: 200*1024)
        killSignal String (Default: 'SIGTERM')

    callback(error, stdout, stderr): 
       On success: error will be null.
       On error: error.code will be the exit code    
    */
    execCmd2: function(cmd, options, callback) {
        exec(cmd, options, function (error, stdout, stderr) {
            callback(error, stdout, stderr);
        });
    }
};
